src/contract_client.rs

```diff
@@ -668,6 +677,193 @@ impl ModuleDeployHandle {
     }
 }
 
+/// The outcome of invoking (dry-running) a transaction to update a smart
+/// contract instance. The variants describe the two cases of successfully
+/// simulating the transaction and rejecting the transaction due to some
+/// reverts.
+pub enum InvokeContractOutcome {
+    /// The simulation of the transaction was successful.
+    Success(SimulatedTransaction),
+    /// The transaction was rejected due to some reverts.
+    Failure(RejectedTransaction),
+}
+
+/// The `InvokedTransaction` type is an alias for the `ContractUpdateBuilder`
+/// type. This type is used when an invoke (dry-run) of a transaction succeeds.
+/// This type includes a convenient send method to send and execute the
+/// transaction on-chain in a subsequent action. As such, it is a builder to
+/// simplify sending smart contract updates.
+pub type SimulatedTransaction = ContractUpdateBuilder;
+
+/// This type is used when an invoke (dry-run) of a transaction gets rejected.
+pub struct RejectedTransaction {
+    /// The return value of the transaction.
+    pub return_value:   Option<ReturnValue>,
+    /// The reject reason of the transaction.
+    pub reason:         RejectReason,
+    /// An optional human-readable decoded reason for the reject reason.
+    /// This is only available if the reject reason is a smart contract logical
+    /// revert and a valid error schema is available for decoding, or the
+    /// reject reason originates from the `concordium-std` crate.
+    pub decoded_reason: Option<String>,
```

Comment:
**Discussion:** I don't think using a string here is ideal, since it is a bad experience if you want to pattern match this. But I don't really have a clear alternative suggestion ready.

Reply:
I return a `serde_json::Value` now instead of a `String`. I also decided to convert the error as decoded from the `concordium-std` crate into the same format as the `error_schema` errors decoded from the smart contract. This way it becomes now abstracted away to developers using the SDK that the errors have different origins.

```
            // Step 1: Try to decode the `reject_reason` using the `concordium-std`
            // error codes.
            if let Some(decoded_error) = decode_concordium_std_error(*reject_reason_code) {
                // Convert the `decoded_error` into the same format as errors decoded
                // with the error schema from smart contracts.
                let mut error_schema_format: HashMap<String, Value> = HashMap::new();
                error_schema_format.insert(decoded_error.to_string(), Value::Array(vec![]));
                return Some(json!(error_schema_format));
            }
```